libia2/ia2.c

```diff
@@ -341,8 +353,14 @@ int protect_pages(struct dl_phdr_info *info, size_t size, void *data) {
       }
 
       if (cur_end > start) {
+        for (size_t i = 0; i < cur_end - start; i += PAGE_SIZE) {
+          volatile char* cur = (volatile char*)start + i;
+          (void)*cur;
+        }
```

Comment:
This should be commented to explain that it probes each page so that we only go on to `pkey_mprotect` regions that we can successfully access all of.

I had some misgivings about whether `volatile` accesses could be reordered with system calls but as I understand it they can't, though I don't actually know how the compiler knows which functions *are* system calls, and I do know that `volatile` accesses can be reordered w/r/t non-dependent non-volatile accesses (and presumably "pure" function calls that are not system calls).